src/test/librbd/test_librbd.cc

```diff
@@ -981,12 +1030,64 @@ TEST_F(TestLibRBD, TestCopyPP)
 
     ASSERT_EQ(0, create_image_pp(rbd, ioctx, name.c_str(), size, &order));
     ASSERT_EQ(0, rbd.open(ioctx, image, name.c_str(), NULL));
+
+    std::string key;
+    std::string val;
+    for (int i = 1; i <= 70; i++) {
+      key = "key" + stringify(i);
+      val = "value" + stringify(i);
+
+      ASSERT_EQ(0, image.metadata_set(key, val));
+    }
+
     ASSERT_EQ(1, test_ls_pp(rbd, ioctx, 1, name.c_str()));
     ASSERT_EQ(0, image.copy(ioctx, name2.c_str()));
     ASSERT_EQ(2, test_ls_pp(rbd, ioctx, 2, name.c_str(), name2.c_str()));
+
+    ASSERT_EQ(0, rbd.open(ioctx, image2, name2.c_str(), NULL));
+
+    const uint32_t MAX_KEYS = 64;
+    map<string, bufferlist> pairs;
+    size_t pairs_size = 0;
+    bool more_results = true;
+    std::string last_key = "";
+    std::string value;
+
+    while (more_results) {
+      ASSERT_EQ(0, image2.metadata_list(last_key, 0, &pairs));
```

Comment:
You should be able to list all 70 keys in this test in one invocation (and below)

Reply:
Can we? we are  iterating over metadata items when listing. just like 
[https://github.com/ceph/ceph/blob/master/src/tools/rbd/action/ImageMeta.cc#L48](url)

Reply:
Indeed -- there is no max restriction on the metadata_list librbd API methods. If such a restriction were added, it could be added to ```librbd::metadata_list``` in ```internal.cc``` so that (1) the OSDs cannot be overloaded,(2) the C and C++ APIs are consistent, and (3) the API-based iteration would only need to be implemented once behind the scene.

Reply:
that's fine. so what about i fix the 'metadata lost' bug in 'internal.cc' to get all metadatas and remove iterations in 'ImageMeta.cc' , 'export.cc' and so on in another pr?  If so, there's no need to use iterations in this pr

Reply:
oh, i misunderstood what you mean. will update soon